src/Storages/ObjectStorage/DataLakes/IcebergMetadata.cpp

```diff
@@ -235,91 +208,230 @@ DataTypePtr getFieldType(const Poco::JSON::Object::Ptr & field, const String & t
+                const ActionsDAG::Node * node = old_node;
+                if (old_type == new_type)
+                {
+                    if (old_json->getValue<String>("name") != name)
+                    {
+                        node = &dag->addAlias(*old_node, name);
+                    }
+                }
+                else if (allowPrimitiveTypeConversion(old_type, new_type))
+                {
+                    node = &dag->addCast(*old_node, getSimpleType(new_type), name);
+                }
+                outputs.push_back(node);
+            }
         }
         else
         {
-            if (schemas->size() != 1)
-                throw Exception(ErrorCodes::UNSUPPORTED_METHOD,
-                    "Cannot read Iceberg table: the table schema has been changed at least 1 time, reading tables with evolved schema is "
-                    "supported. If you want to ignore schema evolution and read all files using latest schema saved on table creation, enable setting "
-                    "iceberg_engine_ignore_schema_evolution (Note: enabling this setting can lead to incorrect result)");
-
-            /// Now we sure that there is only one schema.
-            schema = schemas->getObject(0);
-            if (schema->getValue<int>("schema-id") != current_schema_id)
-                throw Exception(ErrorCodes::BAD_ARGUMENTS, R"(Field "schema-id" of the schema doesn't match "current-schema-id" in metadata)");
+            if (field->isObject("type"))
+            {
+                throw Exception(
+                    ErrorCodes::UNSUPPORTED_METHOD,
+                    "Adding a default column with id {} and complex type is not supported yet, old schema id is {}, new schema id is {}",
+                    id,
+                    current_old_id,
+                    current_new_id);
+            }
+            if (!type->isNullable())
+            {
+                throw Exception(
+                    ErrorCodes::BAD_ARGUMENTS,
+                    "Canßnot add a column with id {} with required values to the table during schema evolution, old schema id is {}, new "
+                    "schema id is {}",
+                    id,
+                    current_old_id,
+                    current_new_id);
+            }
+            ColumnPtr default_type_column = type->createColumnConstWithDefaultValue(0);
+            const auto & constant = dag->addColumn({default_type_column, type, name});
+            outputs.push_back(&constant);
         }
     }
-    else
+    return dag;
+}
+
+std::shared_ptr<const ActionsDAG> IcebergSchemaProcessor::getSchemaTransformationDagByIds(Int32 old_id, Int32 new_id)
+{
+    current_old_id = old_id;
+    current_new_id = new_id;
+    SCOPE_EXIT({
+        current_old_id = -1;
+        current_new_id = -1;
+    });
+    Poco::JSON::Object::Ptr old_schema, new_schema;
+    if (transform_dags_by_ids.contains({old_id, new_id}))
```

Comment:
Please, use `find(key)` method instead of `contains` when you need to access the object. Also use `find(key)` instead of `.at(key)` with `try/catch`